torch/csrc/autograd/profiler_python.cpp

```diff
@@ -0,0 +1,724 @@
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(t <= std::numeric_limits<uint32_t>::max());
+    }
+
+    RawEvent(TraceTag tag, int lasti, TraceContext* ctx, PyCodeObject* f_code)
+            : RawEvent(tag, lasti, ctx) {
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(tag == TraceTag::kPy_Call);
+        misc_.f_code_ = f_code;
+    }
+
+    RawEvent(TraceTag tag, int lasti, TraceContext* ctx, PyObject* arg)
+            : RawEvent(tag, lasti, ctx) {
+        TORCH_INTERNAL_ASSERT_DEBUG_ONLY(tag == TraceTag::kC_Call);
+        misc_.arg_ = arg;
+    }
+
+    uint8_t tag_;
+    uint8_t thread_id_;
+    uint16_t lasti_;
+    uint32_t t_;
+    union {
+        // TraceTag::kPy_Call
+        PyCodeObject* f_code_;
+
+        // TraceTag::kC_Call
+        PyObject* arg_;
+
+        // TraceTag::kPy_Return
+        // TraceTag::kC_Return
+        // ** Unused (placeholder) **
+        void* null_;
+    } misc_;
+
+    C10_NODISCARD TraceTag tag() const {
+        return static_cast<TraceTag>(tag_);
+    }
+
+    C10_NODISCARD int lasti() const {
+        // f_lasti is positive, with one exception: CPython intializes frames
+        // with `f_lasti = -1`. We don't want to give up half of the range by
+        // switching to int16_t. So instead we do the fast (underflowing) cast
+        // in the ctor, and rectify the value in this accessor which should
+        // only be called during trace post processing.
+        return lasti_ == std::numeric_limits<uint16_t>::max()
+            ? (int)(-1)
+            : (int)lasti_;
+    }
+};
+
+// Make sure the bit packing that we do in RawEvent actually results in the
+// desired size reduction.
+static_assert(sizeof(RawEvent) <= 16, "RawEvent is too large");
+
+
+// std::hash doesn't have a specialization for pairs so we have to define one.
+// A simple XOR is good enough for our purposes.
+struct hash_pair {
+    template <class T1, class T2>
+    size_t operator() (const std::pair<T1, T2>& pair) const {
+        return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
+    }
+};
+
+
+// ============================================================================
+// == Tracing implementation ==================================================
+// ============================================================================
+constexpr size_t max_py_threads = std::numeric_limits<uint8_t>::max() + 1;
+
+class PythonTracer final {
+ public:
+    // Static methods serve as external interfaces (which expect raw pointers)
+    // and handle forwarding to the singleton.
+    static void call(Command c);
+
+    static int py_profile_fn(
+        PyObject* obj,
+        PyFrameObject* frame,
+        int what,
+        PyObject* arg);
+
+ private:
+    PythonTracer();
+    static PythonTracer& singleton();
+    friend class PyTraceReplay;
+
+    void start(size_t max_threads = max_py_threads);
+    void stop();
+    void clear();
+
+    void record_py_call(TraceContext* ctx, PyFrameObject* frame);
```

Comment:
Does the PyTorch coding standard differ from FB internal one? I.e. recordPyCall vs record_py_ca;;

Reply:
Nah, I think that's just my python roots showing. I'll fix it.